src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,302 @@
+/**
+ * @file methods/ann/layer/multihead_attention.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the MultiheadAttention class.
+ *
+ * @code
+ * @article{NIPS'17,
+ *   author  = {Ashish Vaswani, Llion Jones, Noam Shazeer, Niki Parmar,
+ *              Aidan N. Gomez, Jakob Uszkoreit, Łukasz Kaiser,
+ *              Illia Polosukhin},
+ *   title   = {Attention Is All You Need},
+ *   year    = {2017},
+ *   url     = {http://arxiv.org/abs/1706.03762v5}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/layer/softmax.hpp>
+#include <mlpack/methods/ann/layer/dropout.hpp>
+#include <mlpack/methods/ann/init_rules/glorot_init.hpp>
+#include <mlpack/methods/ann/regularizer/no_regularizer.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Multihead Attention allows the model to jointly attend to information from
+ * different representation subspaces at different positions. With a single
+ * attention head, averaging inhibits this. [arxiv.org:1706.03762v5]
+ *
```

Comment:
Could you write somewhere an information about the input. I think it's to complex. It could be quite useful to describe the way how the query, the key and the value are stored in the input.

Reply:
Done.